File: hibernate-criteria-builder/src/main/java/io/github/robertomike/hefesto/constructors/ConstructSelectImplementation.java
package io.github.robertomike.hefesto.constructors;

import io.github.robertomike.hefesto.actions.Select;
import io.github.robertomike.hefesto.enums.SelectOperator;
import io.github.robertomike.hefesto.exceptions.QueryException;
import io.github.robertomike.hefesto.utils.HibernateUtils;
import io.github.robertomike.hefesto.models.BaseModel;
import jakarta.persistence.criteria.*;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class ConstructSelectImplementation<T extends BaseModel> extends ConstructSelect {
    private Map<String, Join<?, ?>> joins;
    private CriteriaBuilder cb;

    public ConstructSelectImplementation<T> setJoins(Map<String, Join<?, ?>> joins) {
        this.joins = joins;
        return this;
    }

    public void construct(Root<T> root, CriteriaQuery<T> cr, CriteriaBuilder cb) {
        if (isEmpty()) {
            cr.select(root);
            return;
        }

        multiSelect(root, cr, cb);
    }

    public void multiSelect(Root<?> root, CriteriaQuery<?> cr, CriteriaBuilder cb) {
        this.cb = cb;

        List<Selection<?>> selects = new ArrayList<>();

        items.forEach(element -> {
            Selection<?> select = getSelectField(root, element);


            if (element.getAlias() != null) {
                select = select.alias(element.getAlias());
            }

            selects.add(select);
        });

        cr.multiselect(selects.toArray(new Selection[]{}));
    }

    private Expression<?> getSelectField(Root<?> root, Select element) {
        From<?, ?> from = root;
        var field = element.getField();

        if (element.getField().contains(".") && joins.containsKey(field.split(HibernateUtils.DOT_REGEX)[0])) {
            var splitted = field.split(HibernateUtils.DOT_REGEX);
            from = joins.get(splitted[0]);
            field = splitted[1];
        }

        if (field.contains("*")) {
            return from;
        }

        Expression<?> select = HibernateUtils.getFieldFrom(from, field);

        if (element.getOperator() != null) {
            select = getWithFunction(select, element.getOperator());
        }

        return select;
    }

    @SuppressWarnings("unchecked")
    private Expression<?> getWithFunction(Expression<?> select, SelectOperator operator) {
        return switch (operator) {
            case SUM -> cb.sum((Expression<Number>) select);
            case AVG -> cb.avg((Expression<Number>) select);
            case MAX -> cb.max((Expression<Number>) select);
            case MIN -> cb.min((Expression<Number>) select);
            case COUNT -> cb.count(select);
        };
    }

    @SuppressWarnings("unchecked")
    public void constructSubQuery(Root<?> root, Subquery<?> sub) {
        if (isEmpty()) {
            ((Subquery<Object>) sub).select((Expression<Object>) root);
            return;
        }

        if (size() != 1) {
            throw new QueryException("Sub-query must have only one select when is using Where IN operation");
        }

        ((Subquery<Object>) sub).select((Expression<Object>) getSelectField(root, items.get(0)));
    }
}

File: hibernate-criteria-builder/src/main/kotlin/io/github/robertomike/hefesto/constructors/ConstructSelectImplementation.kt
package io.github.robertomike.hefesto.constructors

import io.github.robertomike.hefesto.actions.Select
import io.github.robertomike.hefesto.enums.SelectOperator
import io.github.robertomike.hefesto.exceptions.QueryException
import io.github.robertomike.hefesto.models.BaseModel
import io.github.robertomike.hefesto.utils.HibernateUtils
import jakarta.persistence.criteria.*

/**
 * Criteria API implementation of SELECT clause construction.
 * 
 * Converts SELECT definitions into JPA Criteria API Selection objects.
 * Supports:
 * - Simple field selections
 * - Aggregate functions (COUNT, SUM, AVG, MIN, MAX)
 * - Field aliases for result mapping
 * - Selections from joined entities
 * - DTO projections via constructor expressions
 * 
 * When no selections are specified, defaults to selecting the entire root entity.
 * For DTO projections, creates a tuple or uses constructor expressions.
 */
class ConstructSelectImplementation<T : BaseModel> : ConstructSelect() {
    private var joins: Map<String, Join<*, *>> = emptyMap()
    private lateinit var cb: CriteriaBuilder

    /**
     * Sets the join map for resolving select fields on joined entities.
     *
     * @param joins map of join aliases to Join objects
     * @return this instance for chaining
     */
    fun setJoins(joins: Map<String, Join<*, *>>): ConstructSelectImplementation<T> {
        this.joins = joins
        return this
    }

    /**
     * Constructs the SELECT clause for a standard entity query.
     * If no selections defined, selects the entire root entity.
     *
     * @param root the root entity
     * @param cr the CriteriaQuery to apply selections to
     * @param cb the CriteriaBuilder for creating expressions
     */
    fun construct(root: Root<T>, cr: CriteriaQuery<T>, cb: CriteriaBuilder) {
        if (isEmpty()) {
            cr.select(root)
            return
        }

        multiSelect(root, cr, cb)
    }

    /**
     * Constructs a multi-select query for DTO projections or tuple results.
     * Used when selecting specific fields rather than entire entities.
     *
     * @param root the root entity
     * @param cr the CriteriaQuery to apply selections to
     * @param cb the CriteriaBuilder for creating expressions
     * @param isProjection true if projecting to a DTO class (requires multiselect)
     */
    fun multiSelect(root: Root<*>, cr: CriteriaQuery<*>, cb: CriteriaBuilder, isProjection: Boolean = false) {
        this.cb = cb

        val selects = mutableListOf<Selection<*>>()

        items.forEach { element ->
            var select: Selection<*> = getSelectField(root, element)

            if (element.alias != null) {
                select = select.alias(element.alias)
            }

            selects.add(select)
        }

        // For a single select without alias, try using select() for simple types
        // BUT not when projecting to a DTO, as multiselect() is required for constructor calls
        if (selects.size == 1 && items[0].alias == null && !isProjection) {
            try {
                @Suppress("UNCHECKED_CAST")
                (cr as CriteriaQuery<Any>).select(selects[0] as Selection<Any>)
            } catch (e: Exception) {
                // Fallback to multiselect if select() doesn't work
                cr.multiselect(*selects.toTypedArray())
            }
        } else {
            cr.multiselect(*selects.toTypedArray())
        }
    }

    private fun getSelectField(root: Root<*>, element: Select): Expression<*> {
        var from: From<*, *> = root
        var field = element.field

        if (element.field.contains(".") && joins.containsKey(field.split(HibernateUtils.DOT_REGEX.toRegex())[0])) {
            val splitted = field.split(HibernateUtils.DOT_REGEX.toRegex())
            from = joins[splitted[0]]!!
            field = splitted[1]
        }

        if (field.contains("*")) {
            return from
        }

        var select: Expression<*> = HibernateUtils.getFieldFrom<Any>(from, field)

        if (element.operator != null) {
            select = getWithFunction(select, element.operator!!)
        }

        return select
    }

    @Suppress("UNCHECKED_CAST")
    private fun getWithFunction(select: Expression<*>, operator: SelectOperator): Expression<*> {
        return when (operator) {
            SelectOperator.SUM -> cb.sum(select as Expression<Number>)
            SelectOperator.AVG -> cb.avg(select as Expression<Number>)
            SelectOperator.MAX -> cb.max(select as Expression<Number>)
            SelectOperator.MIN -> cb.min(select as Expression<Number>)
            SelectOperator.COUNT -> cb.count(select)
        }
    }

    @Suppress("UNCHECKED_CAST")
    fun constructSubQuery(root: Root<*>, sub: Subquery<*>) {
        if (isEmpty()) {
            (sub as Subquery<Any>).select(root as Expression<Any>)
            return
        }

        if (size != 1) {
            throw QueryException("Sub-query must have only one select when is using Where IN operation")
        }

        (sub as Subquery<Any>).select(getSelectField(root, items[0]) as Expression<Any>)
    }
}
